# Identitas Diri
---

**Nama**: Hanan Naufal Indratma  
**Cohort ID**: MC008D5Y2239  
**Group**: MC-10


# **Import Library**

Pada tahap ini, akan diimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [26]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score
from sklearn.linear_model import LogisticRegression, RidgeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB, BernoulliNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import (
    RandomForestClassifier, ExtraTreesClassifier, GradientBoostingClassifier,
    HistGradientBoostingClassifier, AdaBoostClassifier
)
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV

# **Memuat Dataset**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [27]:
# Import Data
url = "https://drive.google.com/uc?export=download&id=15HDlXfQfxhdTX1kFWnuyfAch1kv_ukMQ"
data = pd.read_csv(url)

In [28]:
data

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...        25.380          17.33           184.60      2019.0   
1    ...        24.990          23.41           158.80      1956.0   
2    ...        23.570          25.53           152.50      1709.0   
3    ...        14.910          26.50            98.87       567.7   
4    ...        22.540          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
564  ...        25.450          26.40           166.10      2027.0   
565  ...        23.690          38.25           155.00      1731.0   
566  ...        18.980          34.12           126.70      1124.0   
567  ...        25.740          39.42           184.60      1821.0   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0.2216       

In [29]:
df = data.copy()

In [30]:
df.shape

(569, 32)

# **Exploratory Data Analysis (EDA)**

Pada tahap ini, Anda akan melakukan **Exploratory Data Analysis (EDA)** untuk memahami karakteristik dataset. EDA bertujuan untuk:

   - Tinjau jumlah baris dan kolom dalam dataset.  
   - Tinjau jenis data di setiap kolom .
   - Cek duplikasi data dan missing value.
   - Analisis distribusi variabel numerik dengan statistik deskriptif

Tujuan dari EDA adalah untuk memperoleh wawasan awal yang mendalam mengenai data dan menentukan langkah selanjutnya dalam analisis atau pemodelan.

In [31]:
df.shape

(569, 32)

Dataset memiliki baris sebanyak 569 dan kolom sebanyak 32.

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

Berdasarkan pengecekan pada tipe data, diperoleh tipe data id adalah integer, diagnosis adalah object, serta variabel lain memiliki tipe data float.

In [33]:
df.duplicated().sum()

np.int64(0)

Tidak ditemukan data duplikat.

In [34]:
df.isnull().sum().sum()

np.int64(0)

Tidak ditemukan data hilang/*missing value*.

In [35]:
df.describe()

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

Data memiliki skala nilai yang berbeda-beda, sehingga akan lebih bagus jika dilakukan *scaling* atau *normalization*. 

# **Preprocessing**

Pada tahap ini, data preprocessing adalah langkah penting untuk memastikan kualitas data sebelum digunakan dalam model machine learning. Data mentah sering kali mengandung nilai kosong, duplikasi, atau rentang nilai yang tidak konsisten, yang dapat memengaruhi kinerja model. Oleh karena itu, proses ini bertujuan untuk membersihkan dan mempersiapkan data agar analisis berjalan optimal.

Berikut adalah tahapan-tahapan yang akan dilakukan:
1. Drop Kolom Tidak Diperlukan
2. Encoding Variabel Kategorik
3. Standarisasi Fitur Numerik

In [38]:
# Drop kolom id
df.drop(['id'], axis=1, inplace=True)

Kolom id tidak diperlukan pada saat pemodelan, sehingga kolom id akan dibuang dari dataset.

In [39]:
# Encoding Variabel "diagnosis" menggunakan Label Encoder
df['diagnosis'] = df['diagnosis'].map({'M': 1, 'B': 0})

Dilakukan encoding pada variabel diagnosis dengan mengganti nilai M = 1 dan B = 0.

In [40]:
# Normalisasi Fitur Numerik
from sklearn.preprocessing import MinMaxScaler

# numerical_cols adalah kolom dengan tipe data numerik
numeric_cols = df.select_dtypes(include=['float64', 'int64']).columns

scaler = MinMaxScaler()
df[numeric_cols] = scaler.fit_transform(df[numeric_cols])

Normalisasi fitur menggunakan MinMaxScaler() dilakukan pada variabel numerik agar data memiliki nilai maksimum dan minimum yang sama.

# **Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [41]:
df.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1     0.521037      0.022658        0.545989   0.363733   
1          1     0.643144      0.272574        0.615783   0.501591   
2          1     0.601496      0.390260        0.595743   0.449417   
3          1     0.210090      0.360839        0.233501   0.102906   
4          1     0.629893      0.156578        0.630986   0.489290   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0         0.593753          0.792037        0.703140             0.731113   
1         0.289880          0.181768        0.203608             0.348757   
2         0.514309          0.431017        0.462512             0.635686   
3         0.811321          0.811361        0.565604             0.522863   
4         0.430351          0.347893        0.463918             0.518390   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0       0.686364  ...      0.620776       0.141525         0.668310   
1       0.379798  ...      0.606901       0.303571         0.539818   
2       0.509596  ...      0.556386       0.360075         0.508442   
3       0.776263  ...      0.248310       0.385928         0.241347   
4       0.378283  ...      0.519744       0.123934         0.506948   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    0.450698          0.601136           0.619292         0.568610   
1    0.435214          0.347553           0.154563         0.192971   
2    0.374508          0.483590           0.385375         0.359744   
3    0.094008          0.915472           0.814012         0.548642   
4    0.341575          0.437364           0.172415         0.319489   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              0.912027        0.598462                 0.418864  
1              0.639175        0.233590                 0.222878  
2              0.835052        0.403706                 0.213433  
3              0.884880        1.000000                 0.773711  
4              0.558419        0.157500                 0.142595  

[5 rows x 31 columns]

In [42]:
# Splitting Data df
X = df.drop(['diagnosis'], axis=1)
y = df['diagnosis']

# Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Data splitting dilakukan dengan perbandingan train dan test sebesar 80 : 20.

# **Pemodelan Klasifikasi**


## **a. Membangun Model Klasifikasi**

In [43]:
# Daftar model
models = {
    "Ridge Classifier": RidgeClassifier(),
    "Support Vector Machine": SVC(),
    "Random Forest": RandomForestClassifier(),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric="mlogloss")
}

# Simpan hasil evaluasi
results = []

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average="macro")  # Ubah ke "macro" untuk multiclass
    precision = precision_score(y_test, y_pred, average="macro")
    recall = recall_score(y_test, y_pred, average="macro")

    results.append([name, accuracy, f1, precision, recall])

/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [15:14:06] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


### Penjelasan Algoritma Klasifikasi

Berikut adalah deskripsi dari algoritma-algoritma klasifikasi yang digunakan:

---

#### 1. Ridge Classifier
Ridge Classifier merupakan versi klasifikasi dari **Ridge Regression**, yang menggunakan regularisasi L2 untuk menghindari overfitting. Cocok digunakan pada data multikolinearitas (fitur yang saling berkorelasi).

---

#### 2. Support Vector Machine (SVM)
SVM bekerja dengan mencari hyperplane terbaik yang memisahkan kelas-kelas dalam data. Cocok untuk data berdimensi tinggi dan bisa digunakan dengan kernel (misalnya RBF) untuk menangani data non-linier.

---

#### 3. Random Forest
Ensemble dari banyak Decision Tree yang dibangun pada subset acak data dan fitur. Mengurangi overfitting dan meningkatkan akurasi dengan teknik **bagging**.

---

#### 4. XGBoost
**Extreme Gradient Boosting**, versi cepat dan efisien dari Gradient Boosting. Mendukung regularisasi, early stopping, dan paralelisme. Sangat populer dalam kompetisi data science karena akurasi tinggi dan fleksibilitas.



## **b. Evaluasi Model Klasifikasi**

In [45]:
# Tampilkan hasil dalam dataframe
df_results = pd.DataFrame(results, columns=["Model", "Accuracy", "F1 Score", "Precision", "Recall"])
df_results.sort_values(by="Accuracy", ascending=False, inplace=True)

print(df_results)

                    Model  Accuracy  F1 Score  Precision    Recall
1  Support Vector Machine  0.973684  0.971863   0.974206  0.969702
2           Random Forest  0.964912  0.962302   0.967257  0.958074
0        Ridge Classifier  0.956140  0.952638   0.960473  0.946446
3                 XGBoost  0.956140  0.953106   0.955357  0.951032


- **Support Vector Machine (SVM)** memiliki skor tertinggi untuk semua metrik: akurasi (97.36%), F1 Score (97.18%), precision (97.42%), dan recall (96.97%). Ini menunjukkan bahwa SVM memberikan performa terbaik dalam hal klasifikasi secara keseluruhan.
- **Random Forest** juga menunjukkan performa yang tinggi, namun sedikit di bawah SVM pada setiap metrik.
- **Ridge Classifier dan XGBoost** menunjukkan hasil yang mirip, namun masih kalah dari SVM dan Random Forest.